### PR TITLE
Jason hours log

### LINKED ISSUE #87: Volunteer Hours Card
**Description**
Your task is to create a card on the dashboard for users to be able to view the total hours that they have volunteered, there are two portions to this, one is a simple card with just the total hours volunteered, and the next is a backside to the card with events volunteered and respective hours and date to the right of the event title.

**Acceptance Criteria**
Correctly retrieves sum of events volunteered, and actual events.

Steps:
- Create a VolunteerCard component; follow a similar format to flipped card component [here](https://www.w3schools.com/howto/howto_css_flip_card.asp).
- Get clerk-db id via useUser and fetching relevant data using `/api/user/[userId]`. More detail on this/ impelmentation can be found in the homepage of the react project (page.tsx).
- Parse the fetched data; simple for loop/reducer for total hours volunteered, I recommend fetching all events once via `/api/events,` and using that when going through the `eventsAttended` list, instead of continually calling `/api/events/[eventId]`. 
- Now just list out the eventsAttended, again with a map, I recommend using the [Chakra Table](https://chakra-ui.com/docs/components/table), for easier formatting.

**Additional Information**
image
image
IMG_4171

### PR DESCRIPTION
**Pull request by SilveerDusk**
## Developer: Jason Jelincic

Closes #87 

### Pull Request Summary

Basically completes the hours log screens for both users and admins

### Modifications

/admin/hours.tsx
/hours.tsx
/lib/dates.tsx
/lib/hours.tsx

### Testing Considerations

Window resizing could be is near perfect and date filters work as expected

Videos are the old ones but it is 2am so that can be demoed later

### Pull Request Checklist

- [ ] Code is neat, readable, and works
- [ ] Comments are appropriate
- [ ] The commit messages follows our [guidelines](https://h4i.notion.site/Conventional-Commits-593452ad1179489399ad3bd696ef772a)
- [ ] The developer name is specified
- [ ] The summary is completed
- [ ] Assign reviewers

### Screenshots/Screencast

https://github.com/hack4impact-calpoly/slo-beaver-brigade/assets/114632851/bcb8663a-35ad-428f-b81b-7f719b548639

https://github.com/hack4impact-calpoly/slo-beaver-brigade/assets/114632851/76fcbf45-2537-4caa-aa31-ce8f0420135f

### CODE REVIEW

File: src/app/admin/hours/page.tsx [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+          (event: any) =>
+          event.volunteerEvent && new Date(event.endTime) <= new Date(endDateTime) && new Date(event.startTime) >= new Date(startDateTime)
+          );
+
+        const time = pastEvents.map(
+          (event: any) =>
+            getDuration(event.startTime, event.endTime) *
+            event.attendeeIds.length
+        );
+        setTotalTime(time.reduce((a, b) => a + b, 0));
```
Comment (javalosr2004) [inline src/app/admin/hours/page.tsx:126]: needs typing - int

File: tsconfig.json [hunk 2]
```diff
@@ -26,6 +26,12 @@
       "@database/*": ["database/*"]
     }
   },
-  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
+  "include": [
+    "next-env.d.ts",
+    "**/*.ts",
+    "**/*.tsx",
+    ".next/types/**/*.ts",
+    "src/app/admin/hours"
```
Comment (javalosr2004) [inline tsconfig.json:34]: ???

**Review by javalosr2004 — CHANGES_REQUESTED** [review]
dates aren't staying persistent, as you said, a really big optimization would be to fetch all the events one time, and use that when filtering by making a copy, another optimization would be to first fetch the events server side (server component) and pass it to the actual hours card component (client component), this could be two components or one depending on how you want to go about it, if you choose to have 2 components it makes it real simple, you just pull out any part of the website that uses useState / useEffect and make it it's own component, then you just fetch the events on page.tsx, removing "use client", and pass the events to the corresponding component.

File: tsconfig.json (hunk 2 above)
Comment (SilveerDusk) [reply tsconfig.json:34]: I dont know man. I think it happened during an npm install, I will just remove it and see what happens.

**Review by javalosr2004 — APPROVED** [review]
Excellent use of reusing functions instead of implementing them continually throughout code, good insight with hard coded values and the lower case issue. Amazing work on the hours card!

**Pull request merged**